[docs] Gh #5196 describe current limitations (#5428)
DevExpress/testcafe: docs/articles/documentation/guides/basic-guides/assert.md

> The following web page is an example:

Reviewer: Please change Line 86 (above) to:

`An additional timeout is usually added when you perform asynchronous functional tests.`

Author: Done

method if you use API or specify the [assertion-timeout](../../reference/command-line-interface.md#--assertion-timeout-ms) option

> if you run TestCafe from the command line.

Reviewer: Line 134/129:
`The approach described above allows you to create stable tests with fast runtime and reduces the risk of errors.`
Do you mean:

`The solution described above allows you to create stable tests with improved runtime performance. It also reduces the risk of errors.`

Author: Yes, applied

### Smart Assertion Limitations

> Smart assertions auto-retry only works with:

Reviewer: Can you clarify what you mean by "Smart assertions auto-retry"?

Author: It's a feature. Since we don't really use the word, I'd say:

> Smart assertions auto-retry mechanism only works with:

Reviewer 2: The smart assertion's auto-retry mechanism only works with:

Author: Edited

> * Promises returned from [ClientFunctions](../../reference/test-api/clientfunction/README.md),

Reviewer: ```suggestion
* Promises returned from [ClientFunctions](../../reference/test-api/clientfunction/README.md);
```

Author: Done

> * [Selector](../../reference/test-api/selector/README.md) properties,

Reviewer: ```suggestion
* [Selector](../../reference/test-api/selector/README.md) properties;
```

Author: Done

> * [RequestLogger.count](../../reference/test-api/requestlogger/count.md) properties,

Reviewer: ```suggestion
* [RequestLogger.count](../../reference/test-api/requestlogger/count.md) properties;
```

Author: Done

* [RequestLogger.contains](../../reference/test-api/requestlogger/contains.md) properties.

> The auto retry feature does not work with [DOM Node state](../../reference/test-api/domnodestate.md) properties. If you execute the selector as an asynchronous function, its value is immediately resolved, doesn't update and the corresponding test fails:

Reviewer: Is "auto retry" a feature name (e.g., a button name in TestCafe)? If so, does it need to be capitalized (Auto-Retry feature")?

```suggestion
The auto-retry feature does not work with [DOM Node state](../../reference/test-api/domnodestate.md) properties. If you execute the selector as an asynchronous function, its value is immediately resolved, does not update, and the corresponding test fails:
```

Author: It's not really a name. just a reference

> The auto retry feature does not work with [DOM node snapshot](./select-page-elements.md#dom-node-snapshot) properties. If you execute the selector as an asynchronous function, its value is immediately resolved, doesn't update and the corresponding test fails:

Author: fixed

**Example:**

```html
<html>
    <body>
        <div id="btn" onclick="window.setTimeout(() => this.innerText = 'Loading...', 100)">Click me!</div>
    </body>
</html>
```

```js
test('Button click', async t => {
    const btn = Selector('#btn');

    await t
        .click(btn)

>         .expect(await btn.textContent).contains('Loading...'); //fails because selector value doesn't update according to page behavior

Reviewer: ```suggestion
        .expect(await btn.textContent).contains('Loading...'); //fails because the selector value does not update according to page behavior
```

    await t
        .click(btn)

>         .expect(btn()).contains({ innerText: 'Loading...'}); //fails because selector value doesn't update according to page behavior

Reviewer: ```suggestion
        .expect(btn()).contains({ innerText: 'Loading...'}); //fails because the selector value does not update according to page behavior
```

    await t
        .click(btn)

>         .expect(btn.innerText).eql('Loading...'); //passes because promise returned from the selector eventually updates

Reviewer: ```suggestion
        .expect(btn.innerText).eql('Loading...'); //passes because the promise returned from the selector eventually updates
```

method if you use API or specify the [assertion-timeout](../../reference/command-line-interface.md#--assertion-timeout-ms) option
if you run TestCafe from the command line.

### Smart Assertion Limitations

> Smart assertions auto retry feature only works with:

Author: edited

> * Promises returned from [`ClientFunctions`](../../reference/test-api/clientfunction/README.md),

Reviewer 3: Please share a screenshot

> * Promises returned from [`ClientFunctions`](../../reference/test-api/clientfunction/README.md),

Author: @alexphilin Looks something like this
![image](https://user-images.githubusercontent.com/49588154/90425061-d3845400-e0c7-11ea-9ea4-794091b0b0cf.png)

> * Promises returned from [`ClientFunctions`](../../reference/test-api/clientfunction/README.md),

Reviewer 4: What's the difference with `` `code` `` that is not a link? How can you tell it's a link?

> * Promises returned from [`ClientFunctions`](../../reference/test-api/clientfunction/README.md),

Reviewer 4: We didn't address code + link cases in the styles

> * Promises returned from [`ClientFunctions`](../../reference/test-api/clientfunction/README.md),

Author: > What's the difference with `` `code` `` that is not a link?

from my experience, there is none

> * Promises returned from [`ClientFunctions`](../../reference/test-api/clientfunction/README.md),

Author: I removed the backticks

* [`Selector`](../../reference/test-api/selector/README.md) properties,
* [`RequestLogger.count`](../../reference/test-api/requestlogger/count.md) properties,
* [`RequestLogger.contains`](../../reference/test-api/requestlogger/contains.md) properties.

> The auto retry feature does not work with [DOM node shapshot](./select-page-elements.md#dom-node-snapshot) properties. If you execute the selector as an asynchronous function, its value is immediately resolved, doesn't update and the corresponding test fails:

Reviewer 4: Same as above

Author: removed the backticks

**Example:**

```js
test('Button click', async t => {
    const btn = Selector('#btn');

    await t
        .click(btn)
        .expect(await btn.textContent).contains('Loading...'); //fails because selector value doesn't update according to page behavior

    await t
        .click(btn)
        .expect(btn()).contains({ innerText: 'Loading...'}); //fails because selector value doesn't update according to page behavior

    await t
        .click(btn)
        .expect(btn.innerText).eql('Loading...'); //passes because promise returned from the selector eventually updates
});

> ```

Reviewer 4: Can we also show the page code as an example:

```html
<html>
    <body>
        <div id="btn" onclick="window.setTimeout(() => this.innerText = 'Loading...', 5000)">Click me!</div>
    </body>
</html>
```

If I understand this right

Author: edited
